components/sendgrid/actions/list-blocks/list-blocks.js

```diff
@@ -0,0 +1,77 @@
+const validate = require("validate.js");
+const common = require("../common");
+
+module.exports = {
+  ...common,
+  key: "sendgrid-list-blocks",
+  name: "List Blocks",
+  description:
+    "Allows you to list all email addresses that are currently on your blocks list.",
+  version: "0.0.1",
+  type: "action",
+  props: {
+    ...common.props,
+    startTime: {
+      type: "integer",
+      label: "Start Time",
+      description:
+        "The start of the time range when a blocked email was created (inclusive). This is a unix timestamp.",
+      optional: true,
+    },
+    endTime: {
+      type: "integer",
+      label: "End Time",
+      description:
+        "The end of the time range when a blocked email was created (inclusive). This is a unix timestamp.",
+      optional: true,
+    },
+    numberOfBlocks: {
+      type: "integer",
+      label: "Max # of Blocks to Return",
+      description: "Indicates the max number of blocked emails to return.",
+    },
+  },
+  methods: {
+    ...common.methods,
+  },
+  async run() {
+    const constraints = {
+      numberOfBlocks: {
+        type: "integer",
+      },
+    };
+    if (this.startTime) {
+      constraints.startTime = this.getIntegerGtZeroConstraint();
+    }
+    if (this.endTime) {
+      constraints.endTime = {
+        numericality: {
+          onlyInteger: true,
+          greaterThan: this.startTime > 0 ?
+            this.startTime :
+            0,
+          message: "must be positive integer, non zero, greater than `startTime`.",
+        },
+      };
+    }
+    if (this.numberOfBlocks) {
+      constraints.numberOfBlocks = this.getIntegerGtZeroConstraint();
+    }
+    const validationResult = validate(
+      {
+        startTime: this.startTime,
+        endTime: this.endTime,
+        numberOfBlocks: this.numberOfBlocks,
+      },
+      constraints,
+    );
+    this.checkValidationResults(validationResult);
+    const listBlocksEndpoint = "/v3/suppression/blocks";
+    return this.sendgrid.listItems(
+      listBlocksEndpoint,
+      this.startTime,
+      this.endTime,
+      this.numberOfBlocks,
+    );
```

Comment:
To encapsulate knowledge of the API in the [app file](https://github.com/PipedreamHQ/pipedream/blob/9789bd637e61143a5d980b35cff12758fcee5513/docs/docs/components/guidelines/README.md#good-api-usage), I'd suggest moving the `listBlocksEndpoint` (`/v3/suppression/blocks`) to the app file so it can be reused. For example:

- In `sendgrid.app.js`
```js
    async listBlocks(startTime, endTime, maxItems) {
      const listItemsEndpoint = "/v3/suppression/blocks";
      return this.listItems(listItemsEndpoint, startTime, endTime, maxItems);
    },
```

- In `list-blocks.js`
```js
    this.checkValidationResults(validationResult);
    return this.sendgrid.listBlocks(
      this.startTime,
      this.endTime,
      this.numberOfBlocks,
    );
```

Reply:
Nice! This suggestion has been implemented.